src/guest_memory.rs

```diff
@@ -646,15 +654,15 @@ pub trait GuestMemory {
                 Ok(0) => return Ok(total),
                 // made some progress
                 Ok(len) => {
-                    total += len;
-                    if total == count {
-                        break;
-                    }
+                    total = match total.checked_add(len) {
+                        Some(x) if x < count => x,
+                        Some(x) if x == count => return Ok(x),
+                        _ => return Err(Error::CallbackOutOfRange),
+                    };
```

Comment:
What do you think about (untested)?
```suggestion
                    total = total.checked_add(len).ok_or_else(|| Error::CallbackOutOfRange)?;
                    if total == count {
                        break;
                    };
```
I find that slightly easier to tread and it leaves a single (successful) exit point from the function.

Reply:
We can use `ok_or` because we're not calling any function, we're using a constant, so no need to use the `ok_or_else`. Otherwise, I also find the code much easier to read with your proposal @Ablu.

Reply:
I lightly disagree here on the styling, I think this is a nit so we are okay to merge either way?

Reply:
But isnt that just reverting the logic back to what it was before the PR (e.g. without fixing #269)? E.g. we need to return `Error::CallbackOutOfRange` if `total + len > count`, not just if it overflows.

Reply:
> But isnt that just reverting the logic back to what it was before the PR (e.g. without fixing #269)? E.g. we need to return `Error::CallbackOutOfRange` if `total + len > count`, not just if it overflows.

Ah... I misread the ticket and was thinking we wanted to avoid a usize overflow here. Makes sense. Though I still think that `if total >= count` would be easier on my brain.